src/business-logic/services/device-connection-service-implementation.ts

```diff
@@ -0,0 +1,100 @@
+import { DeviceConnectionService } from './interfaces/device-connection-service'
+import { Device } from '../../model/entities/device'
+import { DeviceConnectionFactory } from './interfaces/device-connection-factory'
+import { DeviceConnection } from './interfaces/inewsgateway-device-connection'
+
+type DeviceAggregate = {
+  deviceConnection: DeviceConnection,
+  device: Device
+}
+
+export class DeviceConnectionServiceImplementation implements DeviceConnectionService {
+  private readonly connectedDevices: Map<string, DeviceAggregate> = new Map()
+
+  constructor(private readonly deviceConnectionFactory: DeviceConnectionFactory) {
+  }
+
+  public async createConnection(device: Device): Promise<DeviceConnectionStatus> {
+    if (this.connectedDevices.has(device.id)) throw new DeviceAlreadyConnectedError(device.id)
+    const deviceConnection: DeviceConnection = this.deviceConnectionFactory.createDeviceConnection(device)
+
+    await deviceConnection.connect()
+
+    this.connectedDevices.set(device.id, {deviceConnection, device})
+
+    return DeviceConnectionStatus.CONNECTED
+  }
+
+  public async send(_deviceId: string, _params: string[]): Promise<void> {
+    const conDevice: DeviceAggregate | undefined = this.connectedDevices.get(_deviceId)
+    if(conDevice === undefined) throw new DeviceNotFoundError(_deviceId)
+
+    await conDevice?.deviceConnection.send(_deviceId, _params)
+  }
+
+  public async listen(_deviceId: string, _callback: (data: unknown) => void): Promise<void> {
+    const conDevice: DeviceAggregate | undefined = this.connectedDevices.get(_deviceId)
+    if(conDevice === undefined) throw new DeviceNotFoundError(_deviceId)
+
+    await conDevice?.deviceConnection.listen(_deviceId, _callback)
+  }
+
+  public getConnectionStatusById(deviceId: string): DeviceConnectionStatus {
+    // we don't do thorough testing of both actual connection and the isConnected field since isConnected is an inheritance from CoreDevice and not needed here.
+    return this.connectedDevices.has(deviceId) ? DeviceConnectionStatus.CONNECTED : DeviceConnectionStatus.DISCONNECTED 
+  }
+
+  public async disconnectConnectionById(deviceId: string): Promise<DeviceConnectionStatus> {
+    await this.connectedDevices.get(deviceId)?.deviceConnection.disconnect()
+    return DeviceConnectionStatus.DISCONNECTED
```

Comment:
Same comment as for the createConnection. Is it necessary to return the status? It should be obvious from the context that the device is disconnected if the call resolves successfully :)

Reply:
High context vs low context communication is, incidentally, the biggest challenge for international teams. Speaking about human interaction, of course.

I am a sucker for low-context code, but will adhere to the team standard. 
Just so I know, is this a strict requirement or a matter of taste?

Reply:
I assume you are referring the explicitness of the interface, when you mention that you prefer low-context code? I would like to hear more about that next time we are both at the office :)

I haven't spoken with others on the team about it, so I am not sure if it is just my opinion or a general consensus, but for me this seems like an extra layer added to the contract.

As a user of the interface, I would have to do two checks instead of one to see if the call was successful. I would first have to check if the promise was resolved or rejected. In the case, where it is rejected, I would most likely be provided with an error/exception and from that conclude that the connect attempt failed and infer the connection status to be `DISCONNECTED`.
In the case were it resolved, the returned connection status should be checked if it is `CONNECTED`. In the case where it is not, I know that it is disconnected, but I have no knowledge of why.

By not returning the DeviceConnectionStatus, I only need to check if the promise resolved or rejected. In the case it resolves, I know the connect attempt was succesful and in the case where it was rejected, I am provided with an error of some sort that describes what went wrong.

Reply:
It's no murder mystery why I prefer explicit feedback: As you yourself you have no idea why because my info is not sufficient. The same can be said for reject and resolve. Why did it happen? 

But I can see that typescript(promises really) forces a standard here, and that for our purposes it is immaterial why it failed. It failed. Enough said. You want to know more, read the logs (I guess).

The reason I want explicit feedback is because I was brought up on C++ ActiveX coding, and we provided HResults with a numeric error code (horrible way of doing things, but that's another matter) whenever something went wrong. Being very explicit was part of my early coder DNA.

When I get rejected by the promise I just have to accept my fate. Javascript didn't want me. Sob.

Reply:
I understand why you prefer the explicit feedback/typings, and so do I. I am a big sucker for strictly typed software. One big disadvantage of promises in TypeScript/JavaScript is that the rejected content can't be typed – it can be literally anything. So part of my suggestion is also based on this.
But one thing that can be done with promises, is that when rejecting e.g. the connect call, the rejection can carry a typed error which we can check for (to obtain the needed information).

E.g. in the following code snippet the error is of type unknown when caught, but we can check if it was an exception for a connection timeout or unknown host. These exceptions could also inherit from a base class e.g. `DisconnectedException`, if we don't care about the explicit reason, but just want to know that it failed and is disconnected.

```typescript
class ConnectionTimeoutException { /* implementation */ }
class UnknownHostException {
  public readonly host: string
  /* rest of the implementation */
}

// ... some code that gets the deviceConnection object ..

try {
  await deviceConnection.connect()
  console.log('Successfully connected.')
} catch (error: unknown) {
  if (error instanceof ConnectionTimeoutException) {
    console.log('Connection timed out when trying to connect.')
  } else if (error instanceof UnknownHostException) {
    console.log(`Tried connecting to unknown host ${error.host}.`)
  } else {
    console.log(`Failed connecting for unknown reason: ${error}`)
  }
}
```

Reply:
TODO: Implement this suggestion. (note to self)